### PR TITLE
Support infallible SockAddr creation

### PR DESCRIPTION
**Pull request by KaiJewson**
- I renamed `SockAddr::init` to `SockAddr::try_init`, to support the addition of an infallible `SockAddr::init`.
- `SockAddr::vsock` can now be made to be infallible by using `SockAddr::init`.
- `SockAddr::try_init` now supports a generic error type, not just `io::Error` for additional flexibility.

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
At the moment I don't really want to release a breaking change as v0.4 is only 2 months old, but we keep a list for possible breaking changes/improvements. So let's revert the `init` -> `try_init` name change.

Maybe we should add more traditional `new` function:
```rust
const fn new(storage: sockaddr_storage, len: socklen_t) -> SockAddr;
```

What do you think @KaiJewson?

**Comment by KaiJewson** [comment]
That might be a good idea - would it suffer from the same copying problem `from_raw_parts` had?

**Comment by Thomasdezeeuw** [comment]
> That might be a good idea - would it suffer from the same copying problem `from_raw_parts` had?

The original `from_raw_parts` took pointer to the address storage and explicitly copied it. The function signature I proposed takes it by value, if the function gets inlined (might require a `#[inline]` attribute) then it wouldn't copy, otherwise I think it would be implicitly copied (because `sockaddr_storage` is `Copy`).

**Comment by KaiJewson** [comment]
That's great then, I will switch this PR to add that. Should it be `unsafe fn` or not?

Another thing I was considering is adding `zeroed` + `as_ptr_mut`/`storage_mut` + `len_mut`/`set_len` so that it can be constructed in-place. Or alternatively, a `fn from_inner<T>(inner: T) -> SockAddr` that fills in the `SockAddr` based on the bytes and size of `inner`.

**Comment by Thomasdezeeuw** [comment]
> That's great then, I will switch this PR to add that. Should it be `unsafe fn` or not?

Good question. I think it should be because we depend on the length being correct. And the family must be set correctly as we depend on that in `as_socket`. Two good things to add to the docs as well.

> Another thing I was considering is adding `zeroed` + `as_ptr_mut`/`storage_mut` + `len_mut`/`set_len` so that it can be constructed in-place. Or alternatively, a `fn from_inner<T>(inner: T) -> SockAddr` that fills in the `SockAddr` based on the bytes and size of `inner`.

I think `new` would fill the spot nicely. Could you give an example of where `new`/`init` aren't sufficient?

**Comment by KaiJewson** [comment]
> I think `new` would fill the spot nicely. Could you give an example of where `new`/`init` aren't sufficient?

`new`/`init` are always sufficient, but it's just about having a slightly nicer API. Here is some code using `new`:

```rust
let mut storage: sockaddr_storage = unsafe { mem::zeroed() };
(&mut storage as *mut sockaddr_storage).cast::<sockaddr_in>().write(sockaddr_in {
    sin_family: AF_INET,
    sin_port: 80,
    sin_addr: in_addr { s_addr: 0 },
    sin_zero: [0; 8],
});
let addr = SockAddr::new(storage, mem::size_of::<sockaddr_in>() as socklen_t);
```

And here's the same code with `from_inner`:

```rust
let addr = SockAddr::from_inner(sockaddr_in {
    sin_family: AF_INET,
    sin_port: 80,
    sin_addr: in_addr { s_addr: 0 },
    sin_zero: [0; 8],
});
```

File: src/sockaddr.rs [hunk 1]
```diff
@@ -12,24 +12,68 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    ///     Domain::from(libc::AF_NETLINK),
+    ///     Type::from(libc::SOCK_DGRAM),
+    ///     Some(Protocol::from(libc::NETLINK_GENERIC)),
+    /// )?;
+    /// socket.bind(&address)?;
+    /// # drop(socket);
+    /// # Ok(())
+    /// # }
+    #[doc = "```"]
+    #[must_use]
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:59]: This isn't required.

File: src/sockaddr.rs [hunk 2]
```diff
@@ -12,24 +12,68 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    /// let socket = Socket::new(
+    ///     Domain::from(libc::AF_NETLINK),
+    ///     Type::from(libc::SOCK_DGRAM),
+    ///     Some(Protocol::from(libc::NETLINK_GENERIC)),
+    /// )?;
+    /// socket.bind(&address)?;
+    /// # drop(socket);
+    /// # Ok(())
+    /// # }
+    #[doc = "```"]
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:58]: You can just use:

    /// ```

File: src/sys/unix.rs [hunk 3]
```diff
@@ -444,6 +444,11 @@ impl SockAddr {
     /// Constructs a `SockAddr` with the family `AF_VSOCK` and the provided CID/port.
     ///
     /// This function is only available on Linux.
+    ///
+    /// # Errors
+    ///
+    /// This function can never fail. In a future version of this library it will be made
+    /// infallible.
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:451]: This shouldn't be part of this pr, let's try to keep it small.

File: src/socket.rs [hunk 4]
```diff
@@ -655,6 +655,7 @@ fn set_common_type(ty: Type) -> Type {
 
 /// Set `FD_CLOEXEC` and `NOSIGPIPE` on the `socket` for platforms that need it.
 #[inline(always)]
+#[allow(clippy::unnecessary_wraps)]
```
Comment (Thomasdezeeuw) [inline src/socket.rs:658]: This doesn't seem necessary.

File: src/sockaddr.rs [hunk 5]
```diff
@@ -58,7 +102,7 @@ impl SockAddr {
     /// # drop(address);
     /// # Ok(())
     /// # }
-    /// ```
+    #[doc = "```"]
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:105]: Please revert this.

File: src/sockaddr.rs [hunk 6]
```diff
@@ -12,24 +12,68 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    ///
+    /// # Safety
+    ///
+    /// - The `SockAddr`'s address family must match its contents.
+    /// - The given length of the `SockAddr` must be correct.
+    ///
+    /// # Examples
+    ///
+    #[cfg_attr(target_os = "linux", doc = "```")]
+    #[cfg_attr(not(target_os = "linux"), doc = "```ignore")]
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:36]: Can you change the example to something that works on all OSs? You can use the same one from `Socket::init`.

File: src/sockaddr.rs [hunk 7]
```diff
@@ -12,24 +12,68 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    ///     Type::from(libc::SOCK_DGRAM),
+    ///     Some(Protocol::from(libc::NETLINK_GENERIC)),
+    /// )?;
+    /// socket.bind(&address)?;
+    /// # drop(socket);
+    /// # Ok(())
+    /// # }
+    #[doc = "```"]
+    #[must_use]
+    pub const unsafe fn new(storage: sockaddr_storage, len: socklen_t) -> Self {
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:60]: Nit: Can you use `SockAddr` instead of `Self`, I always use `Self` for builder patterns, which this isn't.

File: src/sockaddr.rs (hunk 2 above)
Comment (KaiJewson) [reply src/sockaddr.rs:58]: I wrote it this way so that syntax highlighting doesn't break in my editor; with the old method, all doc comments below that would be treated as Rust code. Although I just realized that the example itself could have the appropriate `#![cfg(...)]` at the top and use regular triple-backtick on all platforms, so I'll switch to that instead.

File: src/socket.rs (hunk 4 above)
Comment (KaiJewson) [reply src/socket.rs:658]: It prevents that Clippy warning from firing on platforms where the function is a no-op. Should I remove it?

Comment (Thomasdezeeuw) [reply src/socket.rs:658]: That's why it didn't trigger on macOS. In that case you can leave it.

File: src/sockaddr.rs [hunk 8]
```diff
@@ -12,35 +12,88 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    /// - The address family of the socket address must match its contents.
+    /// - The length of the socket address must be correct.
     ///
     /// # Examples
     ///
-    #[cfg_attr(unix, doc = "```")]
-    #[cfg_attr(not(unix), doc = "```ignore")]
+    /// ```
+    /// # fn main() -> std::io::Result<()> {
+    /// # #[cfg(unix)] {
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:91]: That's a neat solution.

File: src/sockaddr.rs [hunk 9]
```diff
@@ -12,35 +12,88 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
...
+    /// use std::mem;
+    /// use std::os::unix::io::AsRawFd;
+    ///
+    /// use socket2::{SockAddr, Socket, Domain, Type};
+    ///
+    /// let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
+    ///
+    /// // Initialise a `SocketAddr` byte calling `getsockname(2)`.
+    /// let mut addr_storage: libc::sockaddr_storage = unsafe { mem::zeroed() };
+    /// let mut len = 0;
```
Comment (Thomasdezeeuw) [inline src/sockaddr.rs:48]: I think this should be `size_of_val(&addr_storage)` as otherwise the OS think it can only write zero bytes, and thus doesn't write the address.

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
Merged as c24bb6820030c4260876be5ab501e98ffb3058fd, 0624146954f609ec03e94202379b8da447c8b669 and 99f508958eb3d2533b38040eaaaa9284a6a84951. Thanks @KaiJewson.
